WalletWasabi.Tests/UnitTests/BlockchainAnalysis/CoinJoinAnonscoreTests.cs

```diff
@@ -239,4 +240,57 @@ public void InputMergePunishmentNoInheritance()
 		// 10 participants, 1 is you, your anonset would be 10 normally and now too:
 		Assert.Equal(10, tx.WalletOutputs.First().HdPubKey.AnonymitySet);
 	}
+
+	public void SiblingCoinjoinDoesntContributeToAnonScore()
+	{
+		var tx1 = new AnalyzedTransaction();
+		tx1.AddForeignInput();
+		tx1.AddWalletInput();
+		var A = tx1.AddForeignOutput();
+		var B = tx1.AddWalletOutput();
+
+		var tx2 = new AnalyzedTransaction();
+		tx2.AddForeignInput(A);
+		tx2.AddWalletInput(B);
+		tx2.AddForeignOutput();
+		var F = tx2.AddWalletOutput();
+
+		tx1.AnalyzeRecursively();
+
+		Assert.Equal(2, B.Anonymity);
+		Assert.Equal(2, F.Anonymity);
+	}
+
+	[Fact]
+	public void EarlierSiblingCoinjoinDoesntContributeToAnonScore()
+	{
+		var tx1 = new AnalyzedTransaction();
+		tx1.AddForeignInput();
+		tx1.AddWalletInput();
+		var A = tx1.AddForeignOutput();
+		var B = tx1.AddWalletOutput();
+
+		var tx2 = new AnalyzedTransaction();
+		tx2.AddForeignInput();
+		tx2.AddWalletInput(B);
+		var C = tx2.AddForeignOutput();
+		var D = tx2.AddWalletOutput();
+		var E = tx2.AddWalletOutput();
+
+		var tx3 = new AnalyzedTransaction();
+		tx3.AddForeignInput(A);
+		tx3.AddForeignInput(C);
+		tx3.AddWalletInput(D);
+		tx3.AddForeignOutput();
+		tx3.AddForeignOutput();
+		tx3.AddForeignOutput();
+		var L = tx3.AddWalletOutput();
+
+		tx3.AnalyzeRecursively();
+
+		Assert.Equal(2, B.Anonymity);
+		Assert.Equal(2, D.Anonymity);
+		Assert.Equal(2, E.Anonymity);
+		Assert.Equal(3, L.Anonymity);
```

Comment:
This doesn't seem right to me. As far as I understand, this picture is what it wants to recreate, right? https://github.com/zkSNACKs/WalletWasabi/pull/7756#issuecomment-1092834955

![image](https://user-images.githubusercontent.com/9156103/167264245-4e9901c4-b5fa-4bef-90de-b6f4112b98f3.png)

If so, then the result should be 4, not 3:
```suggestion
		Assert.Equal(4, L.Anonymity);
```

This brings us back to the same debate here: https://github.com/zkSNACKs/WalletWasabi/pull/7756#issuecomment-1095597383

Reply:
The anonymity set is coming out as 3 because Tx5 has only 3 inputs, or another way to think of it in this case is that the system of transactions as a whole has only 3 inputs. This is a result of the rule that _the anonymity set cannot be larger than the number of inputs_. This rule exists in WalletWasabi's codebase since 2019, it was introduced in 65b3fd776f7aa238912b4241399c87f96bac0d91 as a result of https://github.com/zkSNACKs/WalletWasabi/issues/1225 and there is even a unit test for it `InputSanityCheck`, see below. Since the rule is unrelated to the issue of mixing with siblings, I'd rather we discuss it in a separate issue.

https://github.com/zkSNACKs/WalletWasabi/blob/2c95d893c99b67ac39db5cc52b0d33b2bcccb3dc/WalletWasabi.Tests/UnitTests/BlockchainAnalysis/CoinJoinAnonscoreTests.cs#L187-L198

Reply:
If that'd be the case, then it'd be ok, but I don't think it is:

> The anonymity set is coming out as 3 because Tx5 has only 3 inputs,

The problem is: the input anonymity score is 2 and "the rule that the anonymity set cannot be larger than the number of inputs" applies to the gained anonymity score, not the final anonymity score.

Reply:
> To better illustrate this in the test case we can set the initial anonymity of B (in the diagram) to something like 10. The resulting anonymity of L should then come out as 12.

I implemented this in https://github.com/zkSNACKs/WalletWasabi/pull/7932/commits/3e506ffd4624bf63a7fb74e9640bb29b00c42211.

Does that make sense to you, @nopara73?